### ISSUE TITLE
ERROR: Scalar indexing is disallowed.

### ISSUE THREAD

**Issue by chuymtz** [rafaelorozco/InvNetsBlog#2]
Hello, 

Great notebook! I'm running into an error and was wondering if you could help me understand how to fix it. Im at the step where I want to start training on the sampled banana dist. I moved things to gpu but uncommenting `f = f |> gpu` and `X = X |> gpu`. 

When i run the loop it very quickly fails with the following message.

```
ERROR: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore should be avoided.

If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
to enable scalar iteration globally or for the operations in question.
```

Any thoughts appreciated. Thanks!

**Comment by chuymtz** [comment]
If I follow the recommendation and run `CUDA.allowscalar(true)`, I get a warning 

```
julia> CUDA.allowscalar(true)
┌ Warning: It's not recommended to use allowscalar([true]) to allow scalar indexing.
│ Instead, use `allowscalar() do end` or `@allowscalar` to denote exactly which operations can use scalar operations.
└ @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:188
```

Then i try to run the loop again and get a different error

```
ERROR: MethodError: no method matching unsafe_convert(::Type{Ptr{Float32}}, ::CuPtr{Float32})

```

**Comment by chuymtz** [comment]
It seems that the issue is in the forward passing in the loss function. After the line 

```
X = X_train[:,:,:,idx]
X = X |> gpu
```
Then passing this `X` in the loss function at `Z, logdet = f.forward(X)` gives the error:
```
ERROR: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore should be avoided.

If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
to enable scalar iteration globally or for the operations in question.
```

`X` looks like this

```
julia> X
1×1×2×300 CuArray{Float32, 4, CUDA.Mem.DeviceBuffer}:
[:, :, 1, 1] =
 -0.39756265

[:, :, 2, 1] =
 5.859172

[:, :, 1, 2] =
 -1.147656

[:, :, 2, 2] =
 9.037992

[:, :, 1, 3] =
 0.17182212

[:, :, 2, 3] =
 5.017807

;;;; … 

[:, :, 1, 298] =
 -1.834815

[:, :, 2, 298] =
 16.520882

[:, :, 1, 299] =
 -0.3137774

[:, :, 2, 299] =
 4.165918

[:, :, 1, 300] =
 1.556241

[:, :, 2, 300] =
 12.716898

julia> size(X)
(1, 1, 2, 300)
```

**Comment by rafaelorozco** [comment]
Hey!
Im having some trouble replicating your errors. 

I suspect that there is a version problem with Julia 1.9 and some of the packages. I ran the code successfully using Julia 1.8 and the newest invertiblenetworks version. 

Could you please retry on Julia 1.8?

**Comment by rafaelorozco** [comment]
Also knowing making sure that you are the newest version of invertiblenetworks would help:
-press the "]" key to go into package mode
-then type "status InvertibleNetworks"
Should look similar to this:
```
julia> 

(sratch) pkg> status InvertibleNetworks
Status `~/projects/sratch/Project.toml`
  [b7115f24] InvertibleNetworks v2.2.6
```

**Comment by chuymtz** [comment]
> Hey! Im having some trouble replicating your errors.
> 
> I suspect that there is a version problem with Julia 1.9 and some of the packages. I ran the code successfully using Julia 1.8 and the newest invertiblenetworks version.
> 
> Could you please retry on Julia 1.8?

Thanks for your reply. I was actually attempting this on julia 10. Give me a min i'll try with julia 1.8. Two things 

* invertiblenetworks.jl version = "2.2.6"
* I was able to run the training loop if I dont send the model and data to gpu. 

Let me try again

**Comment by rafaelorozco** [comment]
Oh if you dont need to use the GPU then I wouldnt worry about it at this stage. For small problems, like the one in the notebook, GPU can actually make it run slower.

**Comment by chuymtz** [comment]
I need to be able to run other parts invertiblenetworks.jl on gpu that's why i was trying to get familiar. 

Btw I ran it on julia 1.8 and it ran on gpu no problem. I also did not have to add the allow scalar to be true. 
Do you recommend for projects involving this pkg that I stick with 1.8?

Thanks!

**Comment by rafaelorozco** [comment]
Great!

 yes I recommend you stick to 1.8 for now, I believe there are some wrinkles with 1.9 that need to be ironed out when I have time.

**Comment by chuymtz** [comment]
Thanks for the tip
